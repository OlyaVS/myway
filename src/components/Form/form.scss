@import '../../style/themes.scss';

.form {
  &__body {
    display: flex;
    justify-content: stretch;
    align-items: center;
  }

  &__input {
    flex-grow: 1;
    min-height: 40px;
    margin-right: 5px;
  }

  &__submit {
    position: relative;
    width: 50px;
    height: 40px;
    font-size: 0;
    border-radius: 6px;
    border: 1px solid map-get($form, submit);
    background-color: map-get($form, submit);

    &::before,
    &::after {
      content: '';
      position: absolute;
      top: 17px;
      left: 17px;
      width: 15px;
      height: 2px;
      background-color: map-get($form, submit-pseudo);
    }

    &::before {
      transform: rotate(90deg);
    }

    &::after {
      transform: rotate(180deg);
    }

    &:hover,
    &:focus {
      background-color: map-get($form, submit-hover);
    }

    &:active {
      background-color: map-get($form, submit-active);
      border-color: map-get($form, submit-active);
    }
  }
}
