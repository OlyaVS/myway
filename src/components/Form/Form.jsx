import React from 'react';

import Button from '../Button/Button.jsx';
import SectionHeader from '../SectionHeader/SectionHeader.jsx';
import TextInput from '../TextInput/TextInput.jsx';
import PropTypes from 'prop-types';

import './form.scss';

function Form(props) {
  return (
    <section className={props.className + ` form`}>
      <SectionHeader className="form__title visually-hidden" title="Address form" />
      <form className="form__body" action="" method="post">
        <TextInput
          classNameLabel="form__label visually-hidden"
          classNameInput="form__input"
          id="address"
          name="address"
          label="address"
          placeholder="Enter address"
        />
        <Button className="form__submit" type="submit" title="submit" value="Add" />
      </form>
    </section>
  );
}

Form.propTypes = {
  className: PropTypes.string.isRequired,
};

export default Form;
