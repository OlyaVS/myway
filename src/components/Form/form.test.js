import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import Form from './Form.jsx';

describe('form testing', () => {
  afterEach(cleanup);

  test('renders form with input field and submit button', () => {
    const handleSubmit = jest.fn();

    const { getByTestId, container } = render(<Form handleSubmit={handleSubmit} />);

    const form = getByTestId('address-form');
    const input = getByTestId('form__input');
    const submitButton = getByTestId('form__button');

    expect(form).toBeTruthy();
    expect(form.tagName).toBe('FORM');
    expect(input.tagName).toBe('INPUT');
    expect(submitButton.tagName).toBe('BUTTON');
    expect(container.firstChild).toMatchSnapshot();
  });

  test('handleChange is called when user types address', () => {
    const handleChange = jest.spyOn(Form.prototype, 'handleChange');
    const handleSubmitProp = jest.fn();

    const { getByTestId } = render(<Form handleSubmit={handleSubmitProp} />);
    const input = getByTestId('form__input');

    fireEvent.change(input, {
      target: {
        value: 'O',
      },
    });

    expect(Form.prototype.handleChange).toHaveBeenCalledTimes(1);
    handleChange.mockClear();
  });

  test('handleSubmit is called when form is submitted', () => {
    const handleChange = jest.spyOn(Form.prototype, 'handleChange');
    const handleSubmit = jest.spyOn(Form.prototype, 'handleSubmit');

    const { getByTestId } = render(<Form handleSubmit={jest.fn()} />);
    const form = getByTestId('address-form');
    const input = getByTestId('form__input');

    fireEvent.change(input, {
      target: {
        value: 'Omsk',
      },
    });

    expect(Form.prototype.handleChange).toHaveBeenCalledTimes(1);
    fireEvent.submit(form);

    expect(Form.prototype.handleSubmit).toHaveBeenCalledTimes(1);
    expect(handleSubmit).toHaveBeenCalledTimes(1);
    handleChange.mockClear();
    handleSubmit.mockClear();
  });

  test('handleMissingValue is called when invalid value is submitted', () => {
    const handleMissingValue = jest.spyOn(Form.prototype, 'handleMissingValue');
    const handleSubmitProp = jest.fn();

    const { getByTestId } = render(<Form handleSubmit={handleSubmitProp} />);
    const input = getByTestId('form__input');

    fireEvent.invalid(input);

    expect(Form.prototype.handleMissingValue).toHaveBeenCalledTimes(1);
    handleMissingValue.mockClear();
  });
});
