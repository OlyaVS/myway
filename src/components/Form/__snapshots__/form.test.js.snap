// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form testing renders form with input field and submit button 1`] = `
<form
  action=""
  class="form__body"
  data-testid="address-form"
  method="post"
>
  <label
    class="form__label visually-hidden label"
    for="address"
  >
    address
  </label>
  <input
    class="form__input input "
    data-testid="form__input"
    id="address"
    name="address"
    placeholder="Enter address"
    required=""
    type="text"
    value=""
  />
  <button
    class="form__submit button"
    data-testid="button"
    title="submit"
    type="submit"
  >
    Add
  </button>
</form>
`;
